h2o-py/h2o/utils/shared_utils.py

```diff
@@ -49,6 +49,37 @@ class List(list):
     pass
 
 
+class LookupList(list):
+    """
+    A list implementation optimized for fast lookups.
+    Some code needs both random/indexed access on large lists and do many lookups `elem in my_list` (e.g. in a loop),
+    it is recommended to use this class in that case to avoid forgetting to build or use a set every time we need a lookup. 
+    
+    Note that this list is read-only as we don't want to have to synchronize the backed set used for the lookups.
+    """
+    def __init__(self, l):
+        super(LookupList, self).__init__(l)
+        self.__set = set(self)  # lookup functions backed by a set
+        
+    def __contains__(self, item):
+        return item in self.__set
+    
+    def __setitem__(self, key, value):
+        raise TypeError("LookupList is read-only")
+    
+    def __delitem__(self, key):
+        raise TypeError("LookupList is read-only")
+    
+    def set(self):
+        """
+        use this for arithmetic operations on the elements to avoid confusion.
+        We still want this to behave like a list for the most part, 
+        and this is a slightly faster than building a set from the list itself.
+        """
+        return set(self.__set)
```

Comment:
Why do we need to wrap it again in set?

Reply:
mutability. Don't want the user to modify the original set

Reply:
but set of a set if quickly build because the impl doesn't have to recheck each item, it can just clone it

Reply:
hang on, I can use a frozenset from the start, this one I can return directly